private/pkg/manifest/manifest_test.go

```diff
@@ -0,0 +1,194 @@
+// Copyright 2020-2022 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package manifest
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"testing"
+	"testing/iotest"
+
+	"github.com/bufbuild/buf/private/pkg/storage/storagemem"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"golang.org/x/crypto/sha3"
+)
+
+func Example() {
+	ctx := context.Background()
+	bucket, _ := storagemem.NewReadBucket(
+		map[string][]byte{
+			"foo": []byte("bar"),
+		},
+	)
+	manifest, _ := NewManifestFromBucket(ctx, bucket)
+	digest, _ := manifest.GetDigest("foo")
+	fmt.Printf("digest[:16]: %s\n", digest.Hex()[:16])
+	path, _ := manifest.GetPath(digest)
+	fmt.Printf("path at digest: %s\n", path)
+	// Output:
+	// digest[:16]: a15163728ed24e1c
+	// path at digest: foo
+}
+
+func mkdigest(content []byte) *Digest {
+	hash := sha3.NewShake256()
+	if _, err := hash.Write(content); err != nil {
+		panic(err)
+	}
+	digest := make([]byte, 64)
+	if _, err := hash.Read(digest); err != nil {
+		panic(err)
+	}
+	return NewDigestFromBytes("shake256", digest)
+}
+
+func TestRoundTripManifest(t *testing.T) {
+	// read a manifest using the unmarshaling method
+	null := mkdigest([]byte{})
+	var manifestBuilder bytes.Buffer
+	for i := 0; i < 2; i++ {
+		fmt.Fprintf(&manifestBuilder, "%s  null%d\n", null, i)
+	}
+	manifestContent := manifestBuilder.Bytes()
+	var m Manifest
+	err := m.UnmarshalText(manifestContent)
+	require.NoError(t, err)
+
+	// marshaling the manifest back should produce an identical result
+	retContent, err := m.MarshalText()
+	require.NoError(t, err)
+	assert.Equal(t, manifestContent, retContent, "round trip failed")
+}
+
+func TestEmptyManifest(t *testing.T) {
+	content, err := NewManifest().MarshalText()
+	assert.NoError(t, err)
+	assert.Equal(t, 0, len(content))
+}
+
+func TestAddContent(t *testing.T) {
+	manifest := NewManifest()
+	manifest.AddContent("null", bytes.NewReader(nil))
```

Comment:
"path" is confusing as well because that's what it literally is called in the code. Trying to read with "path is path" isn't great. I went with "null" because it contains nothing. Would it help if it were more abstract, like "foo?"

Reply:
sure, or
```suggestion
	manifest.AddContent("my/foo/path", bytes.NewReader(nil))
```